In [1]:
%run ../functions.ipynb
%run ../pycodes/read_data.py

In [2]:
import os 

import numpy as np
import pywt as wt

from sklearn.decomposition import PCA
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel

In [3]:
pd.options.display.max_rows = 6
pd.options.display.max_columns = 40

## Getting raw data

In [6]:
path_to_read = '../../data/raw/'
files = []

# r=root, d=directories, f = files
for r, d, f in os.walk(path_to_read):
    for file in f:
        if '.DAT' in file:
            files.append(os.path.join(r, file))

In [7]:
df = pandas.DataFrame(columns = ['ID', 'desg','desi','desr','desz', 'Xaxis'])
for i, obj in enumerate(files):
    ID, filters, time = get_df(obj)
    df.loc[i] = [ID, filters['desg'], filters['desi'], filters['desr'], filters['desz'], time]
#df.to_pickle("./df_raw.pkl")

In [9]:
df = pd.read_pickle("../../data/df_raw.pkl")
mask = (df.desg.str.len() > 1) & (df.desi.str.len() > 1) & (df.desr.str.len() > 1) & (df.desz.str.len() > 1) 
df = df.loc[mask]
df

ID                                               desg  \
0      SN554563  [[0.0, 18.98, 10.49], [1.985000000000582, 0.89...   
1      SN454964  [[0.0, 4.359, 6.2], [4.9799999999959255, 33.82...   
2      SN537864  [[0.0, 0.4987, 4.764], [11.969000000004598, 8....   
...         ...                                                ...   
21314   SN51813  [[14.859000000004016, 11.64, 11.89], [30.91799...   
21315  SN599115  [[0.0, 3.229, 7.693], [10.014999999999418, 1.1...   
21316  SN308423  [[29.875, 4.829, 11.89], [53.88699999999517, 9...   

                                                    desi  \
0      [[0.03200000000651926, 8.001, 9.313], [10.9610...   
1      [[0.010999999998603016, 3.629, 6.148], [4.9959...   
2      [[4.105999999999767, 1.357, 3.683], [13.090000...   
...                                                  ...   
21314  [[12.984000000004016, 5.321, 4.986], [22.88599...   
21315  [[0.027000000001862645, 8.224, 5.11], [3.01099...   
21316  [[0.0, 2.278, 4.54], [7.995999999999185, 3.879...   

                                                    desr  \
0      [[2.0, 5.875, 2.431], [10.121000000006461, 6.9...   
1      [[0.004000000000814907, 0.8701, 3.584], [14.21...   
2      [[0.008000000001629815, 3.18, 2.13], [13.07899...   
...                                                  ...   
21314  [[0.0, 1.53, 1.444], [14.86699999999837, 2.602...   
21315  [[1.0, 0.9777, 3.85], [2.0429999999978463, 0.4...   
21316  [[2.069999999999709, 0.06183, 1.853], [23.0239...   

                                                    desz  \
0      [[15.993000000002212, 1.763, 6.543], [29.95700...   
1      [[5.004000000000815, 4.406, 7.695], [24.139999...   
2      [[3.0750000000043656, 1.838, 3.729], [4.137000...   
...                                                  ...   
21314  [[2.014999999999418, 3.934, 3.402], [13.014999...   
21315  [[1.0429999999978463, 2.444, 4.848], [2.085999...   
21316  [[0.01999999999679858, 6.346, 3.061], [30.0159...   

                                                   Xaxis  
0      [0.0, 1.4326868686868988, 2.8653737373737975, ...  
1      [0.0, 1.2527575757575864, 2.505515151515173, 3...  
2      [0.0, 1.1715252525252502, 2.3430505050505004, ...  
...                                                  ...  
21314  [0.0, 0.7759595959595931, 1.5519191919191861, ...  
21315  [0.0, 1.2428181818181518, 2.4856363636363037, ...  
21316  [0.0, 0.9276363636363202, 1.8552727272726404, ...  

[21291 rows x 6 columns]

## Gaussian Process
- nearly 20 hours to run

- mask filter cuts 28 objetcs (21319 - 21291)

In [ ]:
df = pd.read_pickle("../../data/df_raw.pkl")
mask = (df.desg.str.len() > 1) & (df.desi.str.len() > 1) & (df.desr.str.len() > 1) & (df.desz.str.len() > 1) 
df = df.loc[mask]

df['desg_GP'] = df.apply(lambda df: gplc(df.ID, df.desg[:,0],df.desg[:,1],df.desg[:,2]), axis=1) 
#df.to_pickle("df_GP_scikit_desg.pkl")
df['desi_GP'] = df.apply(lambda df: gplc(df.ID, df.desi[:,0],df.desi[:,1],df.desi[:,2]), axis=1) 
#df.to_pickle("df_GP_scikit_desi.pkl")
df['desr_GP'] = df.apply(lambda df: gplc(df.ID, df.desr[:,0],df.desr[:,1],df.desr[:,2]), axis=1) 
#df.to_pickle("df_GP_scikit_desr.pkl")
df['desz_GP'] = df.apply(lambda df: gplc(df.ID, df.desz[:,0],df.desz[:,1],df.desz[:,2]), axis=1) 
#df.to_pickle("df_GP_scikit_desz.pkl")
#df.to_pickle("../../data/df_GP_scikit.pkl")

In [91]:
df['desg_GP'] = gplc(df.ID.values, df.desg_x.values ,df.desg_y.values, df.desg_err.values)

ValueError: operands could not be broadcast together with shapes (25,) (20,) 

In [29]:
df["desg_x"] = df.apply(lambda df: df.desg[:,0], axis = 1)
df["desg_y"] = df.apply(lambda df: df.desg[:,1], axis = 1)
df["desg_err"] = df.apply(lambda df: df.desg[:,2], axis = 1)

In [30]:
df["desi_x"] = df.apply(lambda df: df.desi[:,0], axis = 1)
df["desi_y"] = df.apply(lambda df: df.desi[:,1], axis = 1)
df["desi_err"] = df.apply(lambda df: df.desi[:,2], axis = 1)

In [31]:
df["desr_x"] = df.apply(lambda df: df.desr[:,0], axis = 1)
df["desr_y"] = df.apply(lambda df: df.desr[:,1], axis = 1)
df["desr_err"] = df.apply(lambda df: df.desr[:,2], axis = 1)

In [32]:
df["desz_x"] = df.apply(lambda df: df.desz[:,0], axis = 1)
df["desz_y"] = df.apply(lambda df: df.desz[:,1], axis = 1)
df["desz_err"] = df.apply(lambda df: df.desz[:,2], axis = 1)

In [85]:
df.desg_err

0        [10.49, 3.211, 25.58, 4.772, 2.991, 2.714, 15....
1        [6.2, 18.69, 33.5, 2.656, 2.466, 9.819, 6.917,...
2        [4.764, 5.686, 2.286, 2.677, 4.819, 3.725, 1.3...
                               ...                        
21314    [11.89, 1.657, 5.981, 3.002, 1.859, 4.335, 3.966]
21315    [7.693, 2.522, 2.19, 6.563, 1.489, 4.948, 8.82...
21316                                [11.89, 5.963, 4.326]
Name: desg_err, Length: 21291, dtype: object

In [45]:
len(df.desg_x.values)

21291

In [18]:
type(df.ID.values)

numpy.ndarray

In [49]:
df.desg_x.values[0]

array([  0.   ,   1.985,  10.106,  15.981,  29.926,  37.93 ,  42.914,
        44.078,  50.063,  51.086,  52.082,  53.067,  57.063,  58.078,
        59.106,  60.106,  65.086,  66.075,  67.086,  68.989,  83.032,
        96.918, 104.969, 109.965, 133.821])

In [10]:
gplc

<function __main__.gplc(index, x, y, yerr, npoints=100, n_restarts_optimizer=100)>

In [ ]:
def gplc(index,x, y, yerr, npoints = 100, n_restarts_optimizer = 100):
  
    X = x.reshape(-1, 1)  
    Xmin, Xmax = X.min(), X.max()
    vary = y.var()
    
    i = y > 0
    
    yerr = yerr[i]/y[i]
    y = np.log(y[i])
    X = X[i]

    const = ConstantKernel(vary, constant_value_bounds=(1e-5, 2*vary))
    rbf = RBF(length_scale = .5*Xmax, length_scale_bounds=(1e-05, Xmax))

    kernel = const * rbf
    
    GPR = GaussianProcessRegressor(kernel = kernel, alpha = yerr**2, n_restarts_optimizer = n_restarts_optimizer)
    
    GPR.fit(X, y)
    
    nX = np.linspace(Xmin, Xmax, npoints).reshape(-1, 1)
    
    ny, covy = GPR.predict(nX, return_cov=True)

    expy = np.exp(ny)
    cov = (expy * expy.reshape(-1, 1)) * covy
    # tempo, resultado, erro
    return nX.ravel(), expy, np.sqrt(cov.diagonal())

In [89]:
def gplc(index,x, y, yerr, npoints = 100, n_restarts_optimizer = 100):
    
    
    return GaussianProcessRegressor(kernel = ConstantKernel(y.var(), constant_value_bounds=(1e-5, 2*y.var())) * RBF(length_scale = .5*x.reshape(-1, 1).max(), length_scale_bounds=(1e-05, x.reshape(-1, 1).max())), alpha = (yerr/y)**2, n_restarts_optimizer = n_restarts_optimizer).fit(x.reshape(-1, 1), np.log(y[i])).predict(np.linspace(x.reshape(-1, 1).min(),x.reshape(-1, 1).max(), npoints).reshape(-1, 1), return_cov=True)
    
    
    
    
   
    #X = x.reshape(-1, 1)#df.loc[:, x].values.reshape(-1, 1)
    #y = y#df.loc[:, y].values
    #yerr = yerr#df.loc[:, yerr].values
    
    #Xmin, Xmax = X.min(), X.max()
    #vary = y.var()
    
    #i = y > 0
    
    #yerr = yerr[i]/y[i]
    #y = np.log(y[i])
    #X = X[i]

#     const = ConstantKernel(y.var(), constant_value_bounds=(1e-5, 2*y.var()))
#     rbf = RBF(length_scale = .5*Xmax, length_scale_bounds=(1e-05, Xmax))

    #kernel = ConstantKernel(y.var(), constant_value_bounds=(1e-5, 2*y.var())) * RBF(length_scale = .5*Xmax, length_scale_bounds=(1e-05, Xmax))
    
    #GPR = GaussianProcessRGettingegressor(kernel = ConstantKernel(y.var(), constant_value_bounds=(1e-5, 2*y.var())) * RBF(length_scale = .5*Xmax, length_scale_bounds=(1e-05, Xmax)), alpha = yerr**2, n_restarts_optimizer = n_restarts_optimizer)
    
  #AQUI  #GaussianProcessRegressor(kernel = ConstantKernel(y.var(), constant_value_bounds=(1e-5, 2*y.var())) * RBF(length_scale = .5*x.reshape(-1, 1).max(), length_scale_bounds=(1e-05, x.reshape(-1, 1).max())), alpha = (yerr/y)**2, n_restarts_optimizer = n_restarts_optimizer).fit(x.reshape(-1, 1), np.log(y[i]))
    
    #nX = np.linspace(X.min(),X.max(), npoints).reshape(-1, 1)
    
    #ny, covy = GPR.predict(np.linspace(X.min(),X.max(), npoints).reshape(-1, 1), return_cov=True)

    #expy = np.exp(ny)
    #cov = (expy * expy.reshape(-1, 1)) * covy
    # tempo, resultado, erro
    ##return GPR.predict(np.linspace(x.reshape(-1, 1).min(),x.reshape(-1, 1).max(), npoints).reshape(-1, 1), return_cov=True)
#nX.ravel(), expy, np.sqrt(cov.diagonal())


In [64]:
df.desg_x.values[0].reshape(-1, 1)[-1][-1]

133.82100000000355

In [71]:
df.desg_y.values[0]

array([18.98  ,  0.8934, 17.59  ,  0.2401,  4.116 ,  8.228 ,  1.194 ,
        8.523 ,  8.755 , 11.51  , 21.08  ,  9.162 , 21.55  ,  0.2404,
        8.358 , 23.25  ,  8.185 ,  3.707 ,  8.578 ,  6.385 ,  5.677 ,
        3.77  ,  3.793 , 13.29  ,  4.118 ])

In [90]:
gplc(df.ID.values[0], df.desg_x.values[0] ,df.desg_y.values[0], df.desg_err.values[0])

IndexError: index 1000 is out of bounds for axis 0 with size 25

In [57]:
gplc(df.ID.values[0], df.desg_x.values[0] ,df.desg_y.values[0], df.desg_err.values[0])

(array([  0.        ,   1.35172727,   2.70345455,   4.05518182,
          5.40690909,   6.75863636,   8.11036364,   9.46209091,
         10.81381818,  12.16554545,  13.51727273,  14.869     ,
         16.22072727,  17.57245455,  18.92418182,  20.27590909,
         21.62763636,  22.97936364,  24.33109091,  25.68281818,
         27.03454545,  28.38627273,  29.738     ,  31.08972727,
         32.44145455,  33.79318182,  35.14490909,  36.49663636,
         37.84836364,  39.20009091,  40.55181818,  41.90354545,
         43.25527273,  44.607     ,  45.95872727,  47.31045455,
         48.66218182,  50.01390909,  51.36563636,  52.71736364,
         54.06909091,  55.42081818,  56.77254545,  58.12427273,
         59.476     ,  60.82772727,  62.17945455,  63.53118182,
         64.88290909,  66.23463636,  67.58636364,  68.93809091,
         70.28981818,  71.64154545,  72.99327273,  74.345     ,
         75.69672727,  77.04845455,  78.40018182,  79.75190909,
         81.10363636,  82.45536364,  83.

In [86]:
df

ID                                               desg  \
0      SN554563  [[0.0, 18.98, 10.49], [1.985000000000582, 0.89...   
1      SN454964  [[0.0, 4.359, 6.2], [4.9799999999959255, 33.82...   
2      SN537864  [[0.0, 0.4987, 4.764], [11.969000000004598, 8....   
...         ...                                                ...   
21314   SN51813  [[14.859000000004016, 11.64, 11.89], [30.91799...   
21315  SN599115  [[0.0, 3.229, 7.693], [10.014999999999418, 1.1...   
21316  SN308423  [[29.875, 4.829, 11.89], [53.88699999999517, 9...   

                                                    desi  \
0      [[0.03200000000651926, 8.001, 9.313], [10.9610...   
1      [[0.010999999998603016, 3.629, 6.148], [4.9959...   
2      [[4.105999999999767, 1.357, 3.683], [13.090000...   
...                                                  ...   
21314  [[12.984000000004016, 5.321, 4.986], [22.88599...   
21315  [[0.027000000001862645, 8.224, 5.11], [3.01099...   
21316  [[0.0, 2.278, 4.54], [7.995999999999185, 3.879...   

                                                    desr  \
0      [[2.0, 5.875, 2.431], [10.121000000006461, 6.9...   
1      [[0.004000000000814907, 0.8701, 3.584], [14.21...   
2      [[0.008000000001629815, 3.18, 2.13], [13.07899...   
...                                                  ...   
21314  [[0.0, 1.53, 1.444], [14.86699999999837, 2.602...   
21315  [[1.0, 0.9777, 3.85], [2.0429999999978463, 0.4...   
21316  [[2.069999999999709, 0.06183, 1.853], [23.0239...   

                                                    desz  \
0      [[15.993000000002212, 1.763, 6.543], [29.95700...   
1      [[5.004000000000815, 4.406, 7.695], [24.139999...   
2      [[3.0750000000043656, 1.838, 3.729], [4.137000...   
...                                                  ...   
21314  [[2.014999999999418, 3.934, 3.402], [13.014999...   
21315  [[1.0429999999978463, 2.444, 4.848], [2.085999...   
21316  [[0.01999999999679858, 6.346, 3.061], [30.0159...   

                                                   Xaxis  \
0      [0.0, 1.4326868686868988, 2.8653737373737975, ...   
1      [0.0, 1.2527575757575864, 2.505515151515173, 3...   
2      [0.0, 1.1715252525252502, 2.3430505050505004, ...   
...                                                  ...   
21314  [0.0, 0.7759595959595931, 1.5519191919191861, ...   
21315  [0.0, 1.2428181818181518, 2.4856363636363037, ...   
21316  [0.0, 0.9276363636363202, 1.8552727272726404, ...   

                                                desg_err  \
0      [10.49, 3.211, 25.58, 4.772, 2.991, 2.714, 15....   
1      [6.2, 18.69, 33.5, 2.656, 2.466, 9.819, 6.917,...   
2      [4.764, 5.686, 2.286, 2.677, 4.819, 3.725, 1.3...   
...                                                  ...   
21314  [11.89, 1.657, 5.981, 3.002, 1.859, 4.335, 3.966]   
21315  [7.693, 2.522, 2.19, 6.563, 1.489, 4.948, 8.82...   
21316                              [11.89, 5.963, 4.326]   

                                                  desg_x  \
0      [0.0, 1.985000000000582, 10.105999999999767, 1...   
1      [0.0, 4.9799999999959255, 5.026999999994587, 1...   
2      [0.0, 11.969000000004598, 13.06300000000192, 1...   
...                                                  ...   
21314  [14.859000000004016, 30.917999999997846, 38.87...   
21315  [0.0, 10.014999999999418, 16.023000000001048, ...   
21316     [29.875, 53.88699999999517, 85.83199999999488]   

                                                  desg_y  \
0      [18.98, 0.8934, 17.59, 0.2401, 4.116, 8.228, 1...   
1      [4.359, 33.82, 48.55, 0.5993, 1.132, 8.722, 10...   
2      [0.4987, 8.389, 3.84, 18.51, 19.24, 18.2, 14.9...   
...                                                  ...   
21314  [11.64, 5.629, 17.42, 30.3, 13.32, 10.08, 0.3053]   
21315  [3.229, 1.198, 0.5477, 6.443, 0.9419, 4.429, 3...   
21316                             [4.829, 9.311, 0.7147]   

                                                  desi_x  \
0      [0.032000000006519

## Wavelets

In [6]:
df_GP = pd.read_pickle("../../data/df_GP_scikit.pkl")
df_GP

ID                                               desg  \
0      SN554563  [[0.0, 18.98, 10.49], [1.985000000000582, 0.89...   
1      SN454964  [[0.0, 4.359, 6.2], [4.9799999999959255, 33.82...   
2      SN537864  [[0.0, 0.4987, 4.764], [11.969000000004598, 8....   
3      SN528418  [[0.0, 4.402, 4.787], [2.0, 4.78, 2.24], [11.9...   
4       SN11302  [[0.0, 7.765, 4.769], [2.997000000003027, 3.75...   
...         ...                                                ...   
21312  SN538509  [[2.989000000001397, 2.453, 2.589], [4.0670000...   
21313  SN710827  [[0.0, 11.09, 4.807], [2.0, 14.44, 2.281], [9....   
21314   SN51813  [[14.859000000004016, 11.64, 11.89], [30.91799...   
21315  SN599115  [[0.0, 3.229, 7.693], [10.014999999999418, 1.1...   
21316  SN308423  [[29.875, 4.829, 11.89], [53.88699999999517, 9...   

                                                    desi  \
0      [[0.03200000000651926, 8.001, 9.313], [10.9610...   
1      [[0.010999999998603016, 3.629, 6.148], [4.9959...   
2      [[4.105999999999767, 1.357, 3.683], [13.090000...   
3      [[0.031000000002677552, 1.92, 5.436], [2.15600...   
4      [[0.023999999997613486, 7.797, 3.15], [3.04299...   
...                                                  ...   
21312  [[0.01599999999598367, 16.46, 3.163], [3.03499...   
21313  [[0.031000000002677552, 23.73, 5.457], [2.1560...   
21314  [[12.984000000004016, 5.321, 4.986], [22.88599...   
21315  [[0.027000000001862645, 8.224, 5.11], [3.01099...   
21316  [[0.0, 2.278, 4.54], [7.995999999999185, 3.879...   

                                                    desr  \
0      [[2.0, 5.875, 2.431], [10.121000000006461, 6.9...   
1      [[0.004000000000814907, 0.8701, 3.584], [14.21...   
2      [[0.008000000001629815, 3.18, 2.13], [13.07899...   
3      [[0.01600000000325963, 4.441, 5.335], [10.0, 0...   
4      [[0.008000000001629815, 1.343, 2.13], [4.09000...   
...                                                  ...   
21312  [[0.0, 11.75, 2.145], [2.995999999999185, 12.8...   
21313  [[0.01600000000325963, 18.35, 5.349], [2.13999...   
21314  [[0.0, 1.53, 1.444], [14.86699999999837, 2.602...   
21315  [[1.0, 0.9777, 3.85], [2.0429999999978463, 0.4...   
21316  [[2.069999999999709, 0.06183, 1.853], [23.0239...   

                                                    desz  \
0      [[15.993000000002212, 1.763, 6.543], [29.95700...   
1      [[5.004000000000815, 4.406, 7.695], [24.139999...   
2      [[3.0750000000043656, 1.838, 3.729], [4.137000...   
3      [[0.06199999999807915, 8.656, 4.694], [2.18699...   
4      [[3.0750000000043656, 6.63, 3.731], [4.1370000...   
...                                                  ...   
21312  [[0.03899999999703141, 13.7, 2.497], [3.067000...   
21313  [[0.06199999999807915, 17.6, 4.704], [2.186999...   
21314  [[2.014999999999418, 3.934, 3.402], [13.014999...   
21315  [[1.0429999999978463, 2.444, 4.848], [2.085999...   
21316  [[0.01999999999679858, 6.346, 3.061], [30.0159...   

                                                   Xaxis  \
0      [0.0, 1.4326868686868988, 2.8653737373737975, ...   
1      [0.0, 1.2527575757575864, 2.505515151515173, 3...   
2      [0.0, 1.1715252525252502, 2.3430505050505004, ...   
3      [0.0, 1.5946565656565574, 3.189313131313115, 4...   
4      [0.0, 1.362696969696962, 2.725393939393924, 4....   
...                                                  ...   
21312  [0.0, 1.060888888888873, 2.121777777777746, 3....   
21313  [0.0, 1.0607272727272974, 2.1214545454545948, ...   
21314  [0.0, 0.7759595959595931, 1.5519191919191861, ...   
21315  [0.0, 1.2428181818181518, 2.4856363636363037, ...   
21316  [0.0, 0.9276363636363202, 1.8552727272726404, ...   

                                                 desg_GP  \
0      ([0.0, 1.3517272727273086, 2.7034545454546173,...   
1      ([0.0, 1.1215252525252062, 2.2430505050504124,...   
2      ([0.0, 1.1417777777777784, 2.283555555555557, ...   
3      ([0.0, 1.453161616161665, 2.9063232323233

In [13]:
df_GP['wavelets'] = df_GP.apply(wavelets, axis = 1)
#df_GP.to_pickle("../../data/df_GP_wav.pkl")

## PCA

In [17]:
df_GP_wav = pd.read_pickle("../../data/df_GP_wav.pkl")
df_GP_wav

ID                                               desg  \
0      SN554563  [[0.0, 18.98, 10.49], [1.985000000000582, 0.89...   
1      SN454964  [[0.0, 4.359, 6.2], [4.9799999999959255, 33.82...   
2      SN537864  [[0.0, 0.4987, 4.764], [11.969000000004598, 8....   
3      SN528418  [[0.0, 4.402, 4.787], [2.0, 4.78, 2.24], [11.9...   
4       SN11302  [[0.0, 7.765, 4.769], [2.997000000003027, 3.75...   
...         ...                                                ...   
21312  SN538509  [[2.989000000001397, 2.453, 2.589], [4.0670000...   
21313  SN710827  [[0.0, 11.09, 4.807], [2.0, 14.44, 2.281], [9....   
21314   SN51813  [[14.859000000004016, 11.64, 11.89], [30.91799...   
21315  SN599115  [[0.0, 3.229, 7.693], [10.014999999999418, 1.1...   
21316  SN308423  [[29.875, 4.829, 11.89], [53.88699999999517, 9...   

                                                    desi  \
0      [[0.03200000000651926, 8.001, 9.313], [10.9610...   
1      [[0.010999999998603016, 3.629, 6.148], [4.9959...   
2      [[4.105999999999767, 1.357, 3.683], [13.090000...   
3      [[0.031000000002677552, 1.92, 5.436], [2.15600...   
4      [[0.023999999997613486, 7.797, 3.15], [3.04299...   
...                                                  ...   
21312  [[0.01599999999598367, 16.46, 3.163], [3.03499...   
21313  [[0.031000000002677552, 23.73, 5.457], [2.1560...   
21314  [[12.984000000004016, 5.321, 4.986], [22.88599...   
21315  [[0.027000000001862645, 8.224, 5.11], [3.01099...   
21316  [[0.0, 2.278, 4.54], [7.995999999999185, 3.879...   

                                                    desr  \
0      [[2.0, 5.875, 2.431], [10.121000000006461, 6.9...   
1      [[0.004000000000814907, 0.8701, 3.584], [14.21...   
2      [[0.008000000001629815, 3.18, 2.13], [13.07899...   
3      [[0.01600000000325963, 4.441, 5.335], [10.0, 0...   
4      [[0.008000000001629815, 1.343, 2.13], [4.09000...   
...                                                  ...   
21312  [[0.0, 11.75, 2.145], [2.995999999999185, 12.8...   
21313  [[0.01600000000325963, 18.35, 5.349], [2.13999...   
21314  [[0.0, 1.53, 1.444], [14.86699999999837, 2.602...   
21315  [[1.0, 0.9777, 3.85], [2.0429999999978463, 0.4...   
21316  [[2.069999999999709, 0.06183, 1.853], [23.0239...   

                                                    desz  \
0      [[15.993000000002212, 1.763, 6.543], [29.95700...   
1      [[5.004000000000815, 4.406, 7.695], [24.139999...   
2      [[3.0750000000043656, 1.838, 3.729], [4.137000...   
3      [[0.06199999999807915, 8.656, 4.694], [2.18699...   
4      [[3.0750000000043656, 6.63, 3.731], [4.1370000...   
...                                                  ...   
21312  [[0.03899999999703141, 13.7, 2.497], [3.067000...   
21313  [[0.06199999999807915, 17.6, 4.704], [2.186999...   
21314  [[2.014999999999418, 3.934, 3.402], [13.014999...   
21315  [[1.0429999999978463, 2.444, 4.848], [2.085999...   
21316  [[0.01999999999679858, 6.346, 3.061], [30.0159...   

                                                   Xaxis  \
0      [0.0, 1.4326868686868988, 2.8653737373737975, ...   
1      [0.0, 1.2527575757575864, 2.505515151515173, 3...   
2      [0.0, 1.1715252525252502, 2.3430505050505004, ...   
3      [0.0, 1.5946565656565574, 3.189313131313115, 4...   
4      [0.0, 1.362696969696962, 2.725393939393924, 4....   
...                                                  ...   
21312  [0.0, 1.060888888888873, 2.121777777777746, 3....   
21313  [0.0, 1.0607272727272974, 2.1214545454545948, ...   
21314  [0.0, 0.7759595959595931, 1.5519191919191861, ...   
21315  [0.0, 1.2428181818181518, 2.4856363636363037, ...   
21316  [0.0, 0.9276363636363202, 1.8552727272726404, ...   

                                                 desg_GP  \
0      ([0.0, 1.3517272727273086, 2.7034545454546173,...   
1      ([0.0, 1.1215252525252062, 2.2430505050504124,...   
2      ([0.0, 1.1417777777777784, 2.283555555555557, ...   
3      ([0.0, 1.453161616161665, 2.9063232323233

In [18]:
# TODO
# Da pra dar uma melhorada nesse codigo aqui ne Felipe?
## Ass: Felipe 15/09/2020
s = df_GP_wav['wavelets']
rslt = np.array(s.tolist())
print(type(rslt))
print(rslt.shape)

pca = PCA(n_components=20) #CHANGE HERE THE NUMBER OF COMPONENTS YOU WANT
pca.fit(rslt)  
print(pca.explained_variance_ratio_)
print(pca.singular_values_)
features = pca.fit_transform(rslt)
print(features.shape)

df_features = pd.DataFrame(features, columns=['f1', 'f2','f3', 'f4','f5',
                                              'f6', 'f7','f8', 'f9','f10',
                                              'f11', 'f12','f13', 'f14','f15',
                                              'f16', 'f17','f18', 'f19','f20'])
df_features

<class 'numpy.ndarray'>
(21291, 1600)
[7.75080477e-01 1.01425485e-01 4.61381626e-02 2.05184206e-02
 1.25061675e-02 9.63334368e-03 7.78383899e-03 6.93479216e-03
 3.43213957e-03 3.21499240e-03 2.15082714e-03 1.72901197e-03
 1.40127624e-03 1.14066078e-03 1.10067200e-03 9.20854977e-04
 7.70792006e-04 6.06373927e-04 5.74896303e-04 4.24922283e-04]
[611412.03415977 221174.02953809 149173.17425849  99479.26030735
  77664.52751964  68163.06501515  61271.35792703  57833.21134671
  40685.81814524  39377.71932654  32207.96882184  28877.49939296
  25996.95377096  23455.18368534  23040.3749391   21074.44776892
  19280.98356476  17101.3676223   16651.57591156  14315.7905789 ]
(21291, 20)


f1           f2          f3          f4          f5  \
0     -152.456150   116.958067   77.435307   13.471403  -51.955130   
1     -312.552895     6.881269 -117.104000  115.002821    9.379711   
2     -153.247121   140.551473   35.816624   -1.885185  -26.144828   
3     -599.929119   227.477497   21.354422   -4.192825   18.183494   
4      261.736801  -354.162164  197.847794  106.060450 -148.757143   
...           ...          ...         ...         ...         ...   
21286 -610.902075   316.496196  -51.323189   26.281654   55.307812   
21287 -385.137025   328.857639  -82.409956   -8.577766   75.881878   
21288  476.795587 -1067.956742 -360.601472 -232.028136 -174.974045   
21289 -612.440524   126.730122  -66.678784  -27.158989   27.419128   
21290 -643.285190   111.940726 -111.379727   18.716777   50.676215   

               f6          f7          f8          f9         f10        f11  \
0      -18.168116   64.503646   28.600517  -27.774363   -8.120098 -20.397438   
1     -139.815067   23.009522  -95.772257  -71.053158  -89.069248  -6.423624   
2       70.170131  -56.272975  -69.939925   72.466017  -20.776364  39.708823   
3      -26.685172   63.081845  -44.028596   -6.045128  -34.436630   5.570809   
4     -143.721834    6.856987  151.184004  -30.020039  -95.881641 -20.564531   
...           ...         ...         ...         ...         ...        ...   
21286  -43.629292  119.323734   -3.941033   12.406115   -2.677448 -27.738459   
21287  -25.120229  108.442031   16.081276   39.783495    6.492771   8.104196   
21288 -144.276123 -483.082741  102.094460  126.465358 -183.045258 -47.553118   
21289  -56.102780    8.534456   -1.585473    3.247079  -47.549219 -12.443284   
21290  -82.445315    0.816497  -95.388172  -36.080611  -55.134534  19.161812   

              f12         f13        f14         f15         f16         f17  \
0       20.430515  -69.650398  75.310421  117.447984  -41.083397  -49.527872   
1      -40.945593   38.165859 -17.001822    7.106318  -35.068364  -10.663220   
2       -1.412278   17.577747   3.413035   14.138187  -37.382939   24.262485   
3      -36.611397   11.885965   1.445781   19.149772   18.009616   11.514932   
4       49.183148  -56.571220   9.639244  155.164602  -37.232036  -11.203544   
...           ...         ...        ...         ...         ...         ...   
21286   20.515848   12.370646   8.089545    5.332065   18.043878   18.107453   
21287   -5.271721   39.555147 -18.821726  -40.400195   36.065939    4.211535   
21288 -106.139239 -126.098258  71.861387  109.377995 -149.822946  168.444144   
21289  -11.492505  -14.476083  -9.730008   20.852124  -41.093223    1.236323   
21290   -3.816460   33.418663 -18.363777  -30.355170   14.762046   20.998782   

              f18        f19        f20  
0       19.079251  17.427973   1.111995  
1        2.622633 -12.197107  -2.859250  
2      -31.411977  -7.438057  18.769267  
3      -26.093907   9.569982  27.505480  
4       -8.507173  27.460559 -19.366518  
...           ...        ...        ...  
21286   18.263960  10.762908  -5.738165  
21287  -18.706366   7.470027  20.365780  
21288  182.912008 -49.802557 -94.981275  
21289   26.710109  16.355968 -23.481705  
21290  -16.922452 -37.872519  28.001602  

[21291 rows x 20 columns]

In [28]:
df_GP_wav_PCA20 = pd.concat([df_GP_wav.reset_index().drop(columns = ['index']), df_features], axis = 1, sort = True)
#df_GP_wav_PCA20.to_pickle("../../data/df_GP_wav.pkl")

In [29]:
df_GP_wav_PCA20

ID                                               desg  \
0      SN554563  [[0.0, 18.98, 10.49], [1.985000000000582, 0.89...   
1      SN454964  [[0.0, 4.359, 6.2], [4.9799999999959255, 33.82...   
2      SN537864  [[0.0, 0.4987, 4.764], [11.969000000004598, 8....   
3      SN528418  [[0.0, 4.402, 4.787], [2.0, 4.78, 2.24], [11.9...   
4       SN11302  [[0.0, 7.765, 4.769], [2.997000000003027, 3.75...   
...         ...                                                ...   
21286  SN538509  [[2.989000000001397, 2.453, 2.589], [4.0670000...   
21287  SN710827  [[0.0, 11.09, 4.807], [2.0, 14.44, 2.281], [9....   
21288   SN51813  [[14.859000000004016, 11.64, 11.89], [30.91799...   
21289  SN599115  [[0.0, 3.229, 7.693], [10.014999999999418, 1.1...   
21290  SN308423  [[29.875, 4.829, 11.89], [53.88699999999517, 9...   

                                                    desi  \
0      [[0.03200000000651926, 8.001, 9.313], [10.9610...   
1      [[0.010999999998603016, 3.629, 6.148], [4.9959...   
2      [[4.105999999999767, 1.357, 3.683], [13.090000...   
3      [[0.031000000002677552, 1.92, 5.436], [2.15600...   
4      [[0.023999999997613486, 7.797, 3.15], [3.04299...   
...                                                  ...   
21286  [[0.01599999999598367, 16.46, 3.163], [3.03499...   
21287  [[0.031000000002677552, 23.73, 5.457], [2.1560...   
21288  [[12.984000000004016, 5.321, 4.986], [22.88599...   
21289  [[0.027000000001862645, 8.224, 5.11], [3.01099...   
21290  [[0.0, 2.278, 4.54], [7.995999999999185, 3.879...   

                                                    desr  \
0      [[2.0, 5.875, 2.431], [10.121000000006461, 6.9...   
1      [[0.004000000000814907, 0.8701, 3.584], [14.21...   
2      [[0.008000000001629815, 3.18, 2.13], [13.07899...   
3      [[0.01600000000325963, 4.441, 5.335], [10.0, 0...   
4      [[0.008000000001629815, 1.343, 2.13], [4.09000...   
...                                                  ...   
21286  [[0.0, 11.75, 2.145], [2.995999999999185, 12.8...   
21287  [[0.01600000000325963, 18.35, 5.349], [2.13999...   
21288  [[0.0, 1.53, 1.444], [14.86699999999837, 2.602...   
21289  [[1.0, 0.9777, 3.85], [2.0429999999978463, 0.4...   
21290  [[2.069999999999709, 0.06183, 1.853], [23.0239...   

                                                    desz  \
0      [[15.993000000002212, 1.763, 6.543], [29.95700...   
1      [[5.004000000000815, 4.406, 7.695], [24.139999...   
2      [[3.0750000000043656, 1.838, 3.729], [4.137000...   
3      [[0.06199999999807915, 8.656, 4.694], [2.18699...   
4      [[3.0750000000043656, 6.63, 3.731], [4.1370000...   
...                                                  ...   
21286  [[0.03899999999703141, 13.7, 2.497], [3.067000...   
21287  [[0.06199999999807915, 17.6, 4.704], [2.186999...   
21288  [[2.014999999999418, 3.934, 3.402], [13.014999...   
21289  [[1.0429999999978463, 2.444, 4.848], [2.085999...   
21290  [[0.01999999999679858, 6.346, 3.061], [30.0159...   

                                                   Xaxis  \
0      [0.0, 1.4326868686868988, 2.8653737373737975, ...   
1      [0.0, 1.2527575757575864, 2.505515151515173, 3...   
2      [0.0, 1.1715252525252502, 2.3430505050505004, ...   
3      [0.0, 1.5946565656565574, 3.189313131313115, 4...   
4      [0.0, 1.362696969696962, 2.725393939393924, 4....   
...                                                  ...   
21286  [0.0, 1.060888888888873, 2.121777777777746, 3....   
21287  [0.0, 1.0607272727272974, 2.1214545454545948, ...   
21288  [0.0, 0.7759595959595931, 1.5519191919191861, ...   
21289  [0.0, 1.2428181818181518, 2.4856363636363037, ...   
21290  [0.0, 0.9276363636363202, 1.8552727272726404, ...   

                                                 desg_GP  \
0      ([0.0, 1.3517272727273086, 2.7034545454546173,...   
1      ([0.0, 1.1215252525252062, 2.2430505050504124,...   
2      ([0.0, 1.1417777777777784, 2.283555555555557, ...   
3      ([0.0, 1.453161616161665, 2.9063232323233

## Concat with redshifts and types

In [33]:
df_redshift = pd.read_pickle("../../data/df_redshift.pkl")
df_redshift

type  type_bool  REDSHIFT_SPEC
ID                                     
SN10001    II      False        0.59162
SN100023   II      False        0.79264
SN10005    Ia       True        1.08327
SN10007    Ia       True        0.82331
SN100085   II      False        0.65600
...       ...        ...            ...
SN99871    II      False        0.93819
SN99899    II      False        0.65243
SN99921    II      False        0.69650
SN99946    Ia       True        0.40625
SN9999     II      False        0.54982

[21317 rows x 3 columns]

In [40]:
df = pd.concat([df_GP_wav_PCA20.set_index('ID'), df_redshift], axis = 1 , join = "inner", sort = 'True')
#df.to_pickle("../../data/df.pkl")

In [41]:
df

desg  \
ID                                                            
SN10001   [[0.0, 2.326, 4.792], [2.0, 4.617, 2.249], [9....   
SN100023  [[6.984000000004016, 5.679, 6.387], [7.9920000...   
SN10005   [[10.027000000001863, 4.996, 8.081], [33.94500...   
SN10007   [[0.0, 6.266, 10.48], [10.105999999999767, 17....   
SN100085  [[0.0, 14.24, 18.69], [0.046999999998661224, 1...   
...                                                     ...   
SN99871   [[5.881999999997788, 8.155, 4.491], [12.851000...   
SN99899   [[15.004000000000815, 1.352, 3.403], [53.87900...   
SN99921   [[0.0, 8.327, 4.797], [2.0, 9.834, 2.261], [9....   
SN99946   [[11.961000000002969, 1.526, 5.683], [13.05500...   
SN9999    [[8.942000000002736, 3.615, 5.646], [18.930000...   

                                                       desi  \
ID                                                            
SN10001   [[0.031000000002677552, 8.672, 5.439], [2.1560...   
SN100023  [[7.00800000000163, 4.698, 7.145], [8.02300000...   
SN10005   [[0.015000000006693881, 0.2977, 2.206], [18.94...   
SN10007   [[0.03200000000651926, 7.161, 9.313], [2.01600...   
SN100085  [[9.245999999999185, 2.122, 2.955], [17.238000...   
...                                                     ...   
SN99871   [[7.010999999998603, 9.268, 9.792], [12.976000...   
SN99899   [[0.0, 4.81, 2.212], [29.9059999999954, 0.0751...   
SN99921   [[0.031000000002677552, 22.78, 5.454], [2.1560...   
SN99946   [[0.01599999999598367, 4.95, 3.147], [3.034999...   
SN9999    [[0.0, 11.54, 6.741], [7.9650000000037835, 4.9...   

                                                       desr  \
ID                                                            
SN10001   [[0.01600000000325963, 8.834, 5.339], [2.13999...   
SN100023  [[0.0, 16.04, 17.59], [6.992000000005646, 19.8...   
SN10005   [[0.0, 1.033, 1.325], [13.015000000006694, 0.4...   
SN10007   [[2.0, 1.24, 2.429], [10.121000000006461, 20.0...   
SN100085  [[0.008000000001629815, 0.4534, 9.048], [23.03...   
...                                                     ...   
SN99871   [[0.0, 0.9223, 1.096], [5.894000000000233, 2.5...   
SN99899   [[7.019999999996799, 2.434, 1.445], [15.027999...   
SN99921   [[0.01600000000325963, 26.73, 5.361], [2.13999...   
SN99946   [[0.0, 0.4826, 2.129], [2.995999999999185, 0.7...   
SN9999    [[8.95400000000518, 0.9289, 4.33], [18.9380000...   

                                                       desz  \
ID                                                            
SN10001   [[0.06199999999807915, 6.026, 4.693], [2.18699...   
SN100023  [[4.121000000006461, 18.01, 3.601], [7.0230000...   
SN10005   [[0.03500000000349246, 1.826, 3.029], [48.875,...   
SN10007   [[0.9770000000062282, 0.6586, 5.329], [15.9930...   
SN100085  [[0.024000000004889444, 8.451, 7.696], [10.168...   
...                                                     ...   
SN99871   [[7.0310000000026776, 1.697, 4.835], [14.01900...   
SN99899   [[0.023999999997613486, 3.478, 3.05], [9.03499...   
SN99921   [[0.06199999999807915, 15.19, 4.7], [2.1869999...   
SN99946   [[0.03899999999703141, 1.045, 2.483], [12.0389...   
SN9999    [[7.973000000005413, 3.831, 6.472], [18.954000...   

                                                      Xaxis  \
ID                                                            
SN10001   [0.0, 1.1313535353535435, 2.262707070707087, 3...   
SN100023  [0.0, 1.3437474747475018, 2.6874949494950036, ...   
SN10005   [0.0, 1.0381919191919462, 2.0763838383838924, ...   
SN10007   [0.0, 1.5136565656566285, 3.027313131313257, 4...   
SN100085  [0.0, 1.2024545454545972, 2.4049090909091944, ...   
...                                                     ...   
SN99871   [0.0, 1.3718787878787768, 2.7437575757575536, ...   
SN99899   [0.0, 0.8468686868686516, 1.6937373737373032, ...   
SN99921   [0.0, 1.2923333333333222, 2.5846666666666445, ...   
SN99946   [0.0, 1.3626161616161374, 2.725232323232275, 4...   
SN9999    [0.0, 1.3